content/codewars-python-test-framework-v2.md

```diff
@@ -0,0 +1,298 @@
+
+* Assertions (using the methods detailed below) can actually be used at any level in the groups, or even outise of them (but you're advised to avoid that).
+* Note that the code executed inside on group isn't stopped at the first failed assertion (unlees you use `allow_raise`. See below), so the full test suite is executed each time.
+
+The basic setup for the tests follow this example:
+
+```python
+@test.describe('Fixed Tests')
+def example_tests():
+
+    @test.it('Example Test Case')
+    def example_test_case():
+        test.assert_equals(add(1, 1), 2, 'Optional Message on Failure')
+    
+    @test.it("More tests")
+    def more():
+        for a,b,exp in [(-2,30,28), (42,0,42)]:
+            test.assert_equals(add(a,b), exp)
+            
+    @test.it("Pair tests")
+    def more():
+        for v in range(10):
+            test.assert_equals(add(v,v), 2*v)
+```
+
+Here is the result visible by the user in the output panel:
+
+`TODO: add the image: "docs/content/PythonTestFrameWorkExample.png"  => how?`
+
+
+
+<br>
+<br>
+
+
+
+
+## Tests grouping (blocks)
+
+The groups are created using the following decorators:
+
+```python
+@test.describe(test_name)
+def _():
+
+    @test.it(test_case_name)
+    def _():
+        ...
+         
+    @test.describe(subgroup_name)
+    def _():
+      ...
+
+        @test.it(test_case_subgroup_name)
+        def _():
+            ...
+```
+
+Those decorators automatically run the decorated function to launch the tests. Both print elapsed time of the block at the end, and `describe`s can be nested.
+
+
+* `describe` blocks:
+
+Those blocks always stay depolyed whatever the results of the tests are, menaing that if you put assertions directly in them, they will be visible in the output, as well as any other `it` block(s) it could contain. They are the basic block inwhich you can insert `it` blocks or other nested `describe` blocks if you need to create subdivisions in your test suite.
+
+* `it` blocks:
+
+On the contrary of the `describe` blocks, the `it` blocks are deployed in the output panel only when they contain failed assertions. Oterwise they are shrinked and only the name of the block and the number of assertions are visible.
+
+Note that you _cannot nest_ the `it` blocks.
+
+
+
+<br>
+<br>
+
+
+
+---
+
+
+## Tests interruption
+
+Some of the functions below can accept a named argument `allow_raise=False`.
+
+If you change its value to `True`, the tests contained inside the current block will be interrupted at the first failed test. The executions are then going back to the parent block if it exists and the next part is executed.
+On some computation-wise intensive katas, this may be a good idea to use this feature so that the user doesn't have to wait a long time before getting feedback (or possibly befor etiming out and in that cas, he might never get ny feedback at all, which may be cumbersome).
```

Comment:
```suggestion
On some computation-wise intensive katas, this may be a good idea to use this feature so that the user doesn't have to wait a long time before getting feedback (or possibly before timing out and in that case, they might never get any feedback at all, which may be cumbersome).
```